utils_pytorch.py

```diff
@@ -103,33 +105,41 @@ def export_model(model, path=None, input_shape=(1, 3, 64, 64)):
     """
     path = get_model_path() if path is None else path
     model = deepcopy(model).cpu().eval()
-    if not isinstance(model, torch.jit.ScriptModule):
-        assert input_shape is not None, "`input_shape` must be provided since model is not a " \
-                                        "`ScriptModule`."
-        traced_model = trace(model, torch.zeros(*input_shape))
-    else:
-        traced_model = model
-    torch.jit.save(traced_model, path)
-    return path
-
-
-def import_model(path=None):
+    if use_script_module:
```

Comment:
Before this line, please check if the model is a ScriptModule instance. If it is, forcibly set use_script_module to True.

Reply:
That's fine. I modified my architecture and code so that it works as a `torch.jit.ScriptModule` .
But during evaluation,if I use without gpu it works fine. In case I set the `"gpu": true` in the `aicrowd.json` file evaluation fails.

I went through PyTorch documentation [https://pytorch.org/docs/stable/_modules/torch/jit.html#load](https://pytorch.org/docs/stable/_modules/torch/jit.html#load) which mentions that the `map_location` should be specified during `torch.jit.load` (e.g., 'cpu', 'cuda:0').

Unlike nn.Module, where model.cuda() works even after loading, for a ScriptModule this does not work.

Hence it should be specified in the `import_model` of `utils_pytorch.py` file 

Can you please confirm this ?

Reply:
Hmm, that's weird. The following snippet works for me:

```python
import torch
import torch.nn as nn
import tempfile as tmp

_, fn = tmp.mkstemp()

m = torch.jit.trace(nn.Linear(1, 1), torch.rand(1, 1))
m.save(fn)

m2 = torch.jit.load(fn)
m2 = m2.cuda()

print(type(m2))
print(m2(torch.rand(1, 1).cuda()).device)
```

The output I get is the following: 
```
<class 'torch.jit.ScriptModule'>
cuda:0
```
Can you double check?

Reply:
Thanks for the test code. Let me check and let you know.

Reply:
```
2019-08-01T02:25:27.916889376Z Traceback (most recent call last):
2019-08-01T02:25:27.91693853Z   File "/home/aicrowd/train_pytorch.py", line 948, in <module>
2019-08-01T02:25:27.916948933Z     exp(args)
2019-08-01T02:25:27.916956762Z   File "/home/aicrowd/train_pytorch.py", line 825, in exp
2019-08-01T02:25:27.916964854Z     pyu.export_model(rep_extractor, input_shape=(1, 3, 64, 64), path=model_path)
2019-08-01T02:25:27.916972654Z   File "/home/aicrowd/utils_pytorch.py", line 108, in export_model
2019-08-01T02:25:27.916980422Z     torch.jit.save(traced_model, path)
2019-08-01T02:25:27.916987523Z   File "/srv/conda/envs/notebook/lib/python3.6/site-packages/torch/jit/__init__.py", line 198, in save
2019-08-01T02:25:27.916995705Z     m.save(f, _extra_files=_extra_files)
2019-08-01T02:25:27.917002834Z   File "/srv/conda/envs/notebook/lib/python3.6/site-packages/torch/jit/__init__.py", line 1205, in save
2019-08-01T02:25:27.917010812Z     return self._c.save(*args, **kwargs)
2019-08-01T02:25:27.917025562Z RuntimeError: [enforce fail at inline_container.cc:307] . PytorchStreamWriter failed opening archive.
```
@nasimrahaman ,  I get this error from my submissions.

I have disabled the gpu and using only cpu for loading and saving the model.

When I run on my machine locally everything works fine.

On the submission containers, all my submissions are getting this error with `torch.jit.save/load`.

Hence none of my submissions are going through.

Reply:
This seems like a file IO issue. I'm guessing `model_path` does not exist -- can you try not specifying a `path` in `pyu.export_model(rep_extractor, input_shape=(1, 3, 64, 64), path=model_path)`?

Reply:
I'm using this for model path as used in `utils_pytorch.py`

```
model_path = pyu.get_model_path()
```

So it should take the same model path for saving. So I suspect the problem is not with `model_path`. Even if the `model_path=none`

`pyu.export_model()` internally gets the model path to save.

I'm getting errors only related to `torch.jit.*` since the beginning and none of my submissions are evaluated till now.

Hence my request to save the model as `nn.Module`.